File: library/src/main/java/cn/wandersnail/http/HttpService.java
package cn.wandersnail.http;

import java.util.Map;

import okhttp3.RequestBody;
import okhttp3.ResponseBody;
import retrofit2.Call;
import retrofit2.http.Body;
import retrofit2.http.DELETE;
import retrofit2.http.FieldMap;
import retrofit2.http.FormUrlEncoded;
import retrofit2.http.GET;
import retrofit2.http.HTTP;
import retrofit2.http.HeaderMap;
import retrofit2.http.POST;
import retrofit2.http.QueryMap;
import retrofit2.http.Url;

/**
 * date: 2019/8/23 20:56
 * author: zengfansheng
 */
public interface HttpService {

    @GET
    Call<ResponseBody> get(@Url String url);

    @GET
    Call<ResponseBody> get(@Url String url, @HeaderMap Map<String, String> headers);

    @POST
    Call<ResponseBody> post(@Url String url);

    @POST
    Call<ResponseBody> post(@Url String url, @HeaderMap Map<String, String> headers);

    @POST
    @FormUrlEncoded
    Call<ResponseBody> postForm(@Url String url, @FieldMap Map<String, Object> params);

    @POST
    Call<ResponseBody> post(@Url String url, @Body RequestBody body);
    
    @POST
    Call<ResponseBody> postParamsAndBody(@Url String url, @FieldMap Map<String, Object> params, @Body RequestBody body);

    @POST
    @FormUrlEncoded
    Call<ResponseBody> postForm(@Url String url, @HeaderMap Map<String, String> headers, @FieldMap Map<String, Object> params);

    @POST
    Call<ResponseBody> post(@Url String url, @HeaderMap Map<String, String> headers, @Body RequestBody body);

    @POST
    Call<ResponseBody> post(@Url String url, @HeaderMap Map<String, String> headers, @FieldMap Map<String, Object> params, @Body RequestBody body);

    @DELETE
    Call<ResponseBody> delete(@Url String url);

    @DELETE
    Call<ResponseBody> deleteParams(@Url String url, @QueryMap Map<String, Object> params);

    @HTTP(method = "DELETE", hasBody = true)
    Call<ResponseBody> delete(@Url String url, @Body RequestBody body);

    @HTTP(method = "DELETE", hasBody = true)
    Call<ResponseBody> deleteParamsAndBody(@Url String url, @QueryMap Map<String, Object> params, @Body RequestBody body);

    @DELETE
    Call<ResponseBody> delete(@Url String url, @HeaderMap Map<String, String> headers);

    @DELETE
    Call<ResponseBody> deleteParams(@Url String url, @HeaderMap Map<String, String> headers,
                                        @QueryMap Map<String, Object> params);

    @HTTP(method = "DELETE", hasBody = true)
    Call<ResponseBody> delete(@Url String url, @HeaderMap Map<String, String> headers, @Body RequestBody body);

    @HTTP(method = "DELETE", hasBody = true)
    Call<ResponseBody> deleteParamsAndBody(@Url String url, @HeaderMap Map<String, String> headers,
                                                           @QueryMap Map<String, Object> params, @Body RequestBody body);
}

File: library/src/main/java/com/snail/network/HttpService.kt
package com.snail.network

import io.reactivex.Observable
import okhttp3.RequestBody
import okhttp3.ResponseBody
import retrofit2.Call
import retrofit2.Response
import retrofit2.http.*

/**
 * 
 *
 * date: 2019/2/25 10:03
 * author: zengfansheng
 */
internal interface HttpService {
    @GET
    fun get(@Url url: String): Observable<Response<ResponseBody>>

    @POST
    @FormUrlEncoded
    fun postForm(@Url url: String, @FieldMap map: Map<String, @JvmSuppressWildcards Any>): Observable<Response<ResponseBody>>

    @POST
    @Headers("Content-Type:application/json;charset=utf-8", "Accept:application/json;")
    fun postJson(@Url url: String, @Body body: RequestBody): Observable<Response<ResponseBody>>

    @POST
    fun post(@Url url: String, @Body body: RequestBody): Observable<Response<ResponseBody>>

    @GET
    fun getSync(@Url url: String): Call<ResponseBody>

    @POST
    @FormUrlEncoded
    fun postFormSync(@Url url: String, @FieldMap map: Map<String, @JvmSuppressWildcards Any>): Call<ResponseBody>

    @POST
    @Headers("Content-Type:application/json;charset=utf-8", "Accept:application/json;")
    fun postJsonSync(@Url url: String, @Body body: RequestBody): Call<ResponseBody>

    @POST
    fun postSync(@Url url: String, @Body body: RequestBody): Call<ResponseBody>
}